node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1852 @@
+					}
+				}
+			}
+		};
+	}
+	// condense the peers to the peers with interest on the candidate
+	let peers = state
+		.peer_views
+		.clone()
+		.into_iter()
+		.filter(|(_peer, view)| {
+			// peers view must contain the candidate hash too
+			state
+				.cached_live_candidates_unioned(view.0.iter())
+				.contains_key(&message_id.0)
+		})
+		.map(|(peer, _)| -> PeerId { peer.clone() })
+		.collect::<Vec<_>>();
+
+	let per_candidate = state.per_candidate.entry(message_id.0.clone()).or_default();
+
+	let peers = peers
+		.into_iter()
+		.filter(|peer| {
+			let peer: PeerId = peer.clone();
+			// avoid sending duplicate messages
+			per_candidate
+				.sent_messages
+				.entry(peer)
+				.or_default()
+				.contains(&message_id)
+		})
+		.collect::<Vec<_>>();
+
+	// gossip that message to interested peers
+	send_tracked_gossip_message_to_peers(ctx, per_candidate, peers, message).await
+}
+
+/// The bitfield distribution subsystem.
+pub struct AvailabilityDistributionSubsystem {
+	/// Pointer to a keystore, which is required for determining this nodes validator index.
+	keystore: KeyStorePtr,
+}
+
+impl AvailabilityDistributionSubsystem {
+	/// The protocol identifier for bitfield distribution.
+	const PROTOCOL_ID: ProtocolId = *b"avad";
+
+	/// Number of ancestors to keep around for the relay-chain heads.
+	const K: usize = 3;
+
+	/// Create a new instance of the availability distribution.
+	pub fn new(keystore: KeyStorePtr) -> Self {
+		Self { keystore }
+	}
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(self, mut ctx: Context) -> Result<()>
+	where
+		Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await
+		.map_err::<Error, _>(Into::into)?;
+
+		// work: process incoming messages from the overseer.
+		let mut state = ProtocolState::default();
+		loop {
+			let message = ctx.recv().await.map_err::<Error, _>(Into::into)?;
+			match message {
+				FromOverseer::Communication {
+					msg: AvailabilityDistributionMessage::NetworkBridgeUpdate(event),
+				} => {
+					trace!(target: TARGET, "Processing NetworkMessage");
+					// a network message was received
+					if let Err(e) = handle_network_msg(
+						&mut ctx,
+						self.keystore.clone(),
+						&mut state,
+						event
+					).await {
+						warn!(
+							target: TARGET,
+							"Failed to handle incomming network messages: {:?}", e
+						);
+					}
+				}
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
+					activated,
+					deactivated,
+				})) => {
+					for relay_parent in activated {
+						trace!(target: TARGET, "Start {:?}", relay_parent);
+					}
+					for relay_parent in deactivated {
+						trace!(target: TARGET, "Stop {:?}", relay_parent);
+					}
+				}
```

Comment:
we don't handle this signal here, not sure what's the purpose of logs

Reply:
It's only trace logs, but there was no particular purpose here. Can be removed.